server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java

```diff
@@ -422,27 +434,33 @@ private void onShardOperation(final NodeRequest request, final ShardRouting shar
                 if (logger.isTraceEnabled()) {
                     logger.trace("[{}]  executing operation for shard [{}]", actionName, shardRouting.shortSummary());
                 }
-                shardOperation(request.indicesLevelRequest, shardRouting, listener);
+                // Call below can throw - turning an exception into BroadcastShardOperationFailedException is taken care in this method.
+                shardOperationAndWrapFailure(request.indicesLevelRequest, shardRouting, listener);
```

Comment:
Another option could be keep calling shardOperation but update all subclasses to handle exception - but I found it a bit error prone. We can forget to do that if we (ever) add another subclass.

Reply:
I think this would be neater if we wrapped the listener here with another one that overrides `onFailure` and constructs the exception appropriately before delegating?

Reply:
Thanks, reverted initial commit and used wrapped listener in https://github.com/crate/crate/pull/16967/commits/f367951697b1b701da94d3a057e635e5112a8292

Just to be sure: looks like 3 out 4 TransportBroadcastByNodeAction subclasses were fine on master, if code throws, it hit our catch-and-convert logic. I think the only way to hit ClassCastException was  TransportSyncRetentionLeasesAction -> ... -> [RetentionLeaseSyncAction.sync](https://github.com/crate/crate/blob/master/server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseSyncAction.java#L121-L125)